HIVE-27956: Query based compactor implementation separation

kuczoram opened the pull request:
Refactored the compaction query generation logic. The main idea was to separate the code parts responsibe for different compaction types. So far everything was handled by one class, therefore it was hard to read and changing the logic for one compaction type might caused side effects on other types. 

There was an older PR for this change (https://github.com/apache/hive/pull/4952), but it got far away from the master branch, and it was easier to create a new PR.
In this PR, I addressed all the comments from the previous one.
Also added unit test for the query generation. I ran these tests before the refactoring as well, to be sure that no query output was changed.

zratkai commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderTest.java:
```diff
@@ -0,0 +1,146 @@
+
+import static org.mockito.Mockito.when;
+
+public class CompactionQueryBuilderTest {
```
I do not see any test case in this class, so I the name is misleading. Please rename it to e.g. CompactionQueryBuilderTestingHelper

kuczoram replied:
Thanks for the hint, I renamed it.

zratkai commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -275,137 +250,105 @@ String build() {
-    }
+    deltas.forEach(
+        delta -> query.append("partition (file_name='").append(delta.getPath().getName()).append("')" + " location '")
+            .append(delta.getPath()).append("' "));
```
Please avoid the concatenation : append("')" + " location '")

kuczoram replied:
Fixed it.

zratkai commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForInsertOnly.java:
```diff
@@ -0,0 +1,265 @@
+
+  @Override
+  protected void buildSelectClauseForInsert(StringBuilder query) {
+    switch (compactionType) {
```
The goal of this commit is to split switch like this into classes?
So I think this should be split into: 
CompactionQueryBuilderForMajor and CompactionQueryBuilderForMinor

kuczoram replied:
Yeah, but then we should make the switch in the CompactionQueryBuilderForMajor and CompactionQueryBuilderForMinor classes, to differentiate between insert-only and acid tables. I choose to keep the logic for the insert-only tables in one class. What I can do is split them up to CompactionQueryBuilderForInsertOnlyMajor and CompactionQueryBuilderForInsertOnlyMinor. I am not sure if it would make the design better, but I am absolutely open for discussing it. What do you think?

zratkai replied:
I think this in CompactionQueryBuilderFactory class breaks a little bit of the logic/aim of this commit:

    if (insertOnly) {
      return new CompactionQueryBuilderForInsertOnly(compactionType);
    } else {
      switch (compactionType) {
      case MAJOR:
        return new CompactionQueryBuilderForMajor();
      case MINOR:
        return new CompactionQueryBuilderForMinor();
      case REBALANCE:
        return new CompactionQueryBuilderForRebalance();
      }
    }
    
    I would keep this part and propagate the insertOnly property like:
    
     switch (compactionType) {
      case MAJOR:
        return new CompactionQueryBuilderForMajor(insertOnly);
      case MINOR:
        return new CompactionQueryBuilderForMinor(insertOnly);
      case REBALANCE:
        return new CompactionQueryBuilderForRebalance(insertOnly);
        
        
        and split this switch cases between them accordingly.

kuczoram replied:
I see you point. But the insert-only compaction is kind of a special case. I would like to keep it separately from the ACID compactions. The original aim was to be able to change things in the compaction query generation logic more easily and by not breaking the other types of compactions. In my head, the separation was always like 'MAJOR', 'MINOR' and the compactions for the insert-only tables, because they are kind of specials. I would rather not put the insert-only logic into the acid compactions. I am ok to create separate files for the insert-only MAJOR and MINOR, so we can get rid of this switch case, but I wouldn't like to put the 'if (insertOnly)' choices into the MAJOR and MINOR compaction code.

zratkai replied:
Ok, thank keep as it is in this change.

zratkai reviewed (changes requested):
Please check my comments!

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForMinor.java:
```diff
@@ -0,0 +1,106 @@
+    return bucketingVersion;
+  }
+
+  protected org.apache.hadoop.hive.ql.metadata.Table getTable() throws HiveException {
+    org.apache.hadoop.hive.ql.metadata.Table t = Hive.get().getTable(sourceTab.getDbName(), sourceTab.getTableName());
+    return t;
+  }
```
nit
can we return directly?
```
  protected org.apache.hadoop.hive.ql.metadata.Table getTable() throws HiveException {
    return Hive.get().getTable(sourceTab.getDbName(), sourceTab.getTableName());
```

kuczoram replied:
Sure, fixed it.

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:
```diff
@@ -54,6 +54,8 @@ public abstract class QueryCompactor implements Compactor {
   private static final Logger LOG = LoggerFactory.getLogger(QueryCompactor.class.getName());
   private static final String COMPACTOR_PREFIX = "compactor.";
 
+  protected CompactionQueryBuilderFactory compactionQueryBuilderFactory = new CompactionQueryBuilderFactory();
```
Why is this ``protected``?

kuczoram replied:
Oh, sorry, forgot to send the answer here. I created an instance of the CompactionQueryBuilderFactory here, but it is used in the subclasses, that's why it is protected. But I realized, it is not a very good approach, so I changed it to create the instance where it is used.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderTest.java:
```diff
@@ -0,0 +1,146 @@
+
+  protected ValidCompactorWriteIdList createWriteId(long minWriteId) {
+    long[] abortedWriteIdList = { 1111L };
+    ValidCompactorWriteIdList writeIds =
+        new ValidCompactorWriteIdList("comp_test_source_table", abortedWriteIdList, null, 15L, minWriteId);
+    return writeIds;
```
maybe we can return directly
```
    return new ValidCompactorWriteIdList("comp_test_source_table", abortedWriteIdList, null, 15L, minWriteId);
```

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderTest.java:
```diff
@@ -0,0 +1,146 @@
+        new ValidCompactorWriteIdList("comp_test_source_table", abortedWriteIdList, null, 15L, minWriteId);
+    return writeIds;
+  }
+
```
super nit
avoid this empty line

kuczoram replied:
Removed it.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMajorCompaction.java:
```diff
@@ -0,0 +1,261 @@
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.ql.io.AcidDirectory;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
```
check if the tests are working.
it should be 
```
import org.junit.Test;
```

else your assertEquals should be from Assertions., I think as of now this test has both junit-4 & 5 together

kuczoram replied:
Thanks for pointing this out, cleaned these up. The tests work fine with org.junit.Test

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMajorCompaction.java:
```diff
@@ -0,0 +1,261 @@
+import java.util.Collections;
+
+import static org.junit.Assert.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
Not sure if you need static import, for assertEquals we didn't had, anyway this is junit-5
```
import static org.junit.Assert.assertTrue;
```

kuczoram replied:
We can do it like the assertEquals, like 'Assert.assertTrue'.
I changed these.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMajorCompaction.java:
```diff
@@ -0,0 +1,261 @@
+    queryBuilder.setSourceTab(sourceTable);
+    queryBuilder.setSourcePartition(sourcePartition);
+
+    Exception exception = assertThrows(IllegalStateException.class, () -> {
+      queryBuilder.build();
+    });
```
Can we change to 
```
    Exception exception = assertThrows(IllegalStateException.class, queryBuilder::build);
```

kuczoram replied:
Fixed this.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMinorCompaction.java:
```diff
@@ -0,0 +1,186 @@
+
+public class TestCompactionQueryBuilderForMinorCompaction extends CompactionQueryBuilderTest {
+
+  class CompactionQueryBuilderForMinorMock extends CompactionQueryBuilderForMinor {
```
maybe the class can be static

kuczoram replied:
Sure, I changed it.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMinorCompaction.java:
```diff
@@ -0,0 +1,186 @@
+import org.apache.hadoop.hive.ql.io.AcidDirectory;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
```
use the same as the assert
```
import org.junit.Test;
```

kuczoram replied:
Fixed these.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMmCompaction.java:
```diff
@@ -0,0 +1,475 @@
+import org.apache.hadoop.hive.ql.io.AcidDirectory;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
```
``import org.junit.Test;``

kuczoram replied:
Fixed these.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMmCompaction.java:
```diff
@@ -0,0 +1,475 @@
+import java.util.Map;
+
+import static org.junit.Assert.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
should be
``import static org.junit.Assert.assertTrue;``

kuczoram replied:
Fixed these.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMmCompaction.java:
```diff
@@ -0,0 +1,475 @@
+    queryBuilder.setSourceTab(sourceTable);
+    StorageDescriptor storageDescriptor = createStorageDescriptor();
+    queryBuilder.setStorageDescriptor(storageDescriptor);
+    Exception exception = assertThrows(RuntimeException.class, () -> {
+      queryBuilder.build();
+    });
```
Can replace with 
```
    Exception exception = assertThrows(RuntimeException.class, queryBuilder::build);
```

kuczoram replied:
Thanks for the hint, I replaced these.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForRebalanceCompaction.java:
```diff
@@ -0,0 +1,175 @@
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.ql.io.AcidDirectory;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.Assert.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
Same import issues like above

kuczoram replied:
Fixed these.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForRebalanceCompaction.java:
```diff
@@ -0,0 +1,175 @@
+    Assert.assertEquals(expectedQuery, query);
+  }
+
+  public void testAlter() {
```
Misses test annotation

kuczoram replied:
Thanks, fixed that.

ayushtkn reviewed (commented):
Thanx @kuczoram dropped some comments, in general looks good.

There were some commments from @aturoczy on the previous PR, please check if they are addressed as well

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -17,72 +17,45 @@
-import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
-import org.apache.hadoop.hive.metastore.api.Table;
-import org.apache.hadoop.hive.metastore.api.hive_metastoreConstants;
+import org.apache.hadoop.hive.metastore.api.*;
```
Thanks a lot for pointing this out. I remember removing all of these, but before committing, I did an autoformat on the classes and it seems it added them back and I didn't notice. The Hive style should be set in my IDE, but it seems something went wrong. But anyway, I didn't notice these, so thanks a lot for finding them.

kuczoram replied:
Fixed these in all affected classes.

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -17,72 +17,45 @@
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Fixed these in all affected classes

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderFactory.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.hadoop.hive.metastore.api.CompactionType;
+
+class CompactionQueryBuilderFactory {
+  public CompactionQueryBuilder getCompactionQueryBuilder(CompactionType compactionType,
+      boolean insertOnly){
```
Fixed it.

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -419,265 +362,45 @@ private void buildWhereClauseForInsert(StringBuilder query) {
+    if (sourceTab != null) {
+      query.append("(");
       query.append(
-          "`operation` int, `originalTransaction` bigint, `bucket` int, `rowId` bigint, "
-              + "`currentTransaction` bigint, `row` struct<");
+          "`operation` int, `originalTransaction` bigint, `bucket` int, `rowId` bigint, " + "`currentTransaction` bigint, `row` struct<");
+      List<String> columnDescs = getColumnDescs();
+      query.append(StringUtils.join(columnDescs, ','));
+      query.append(">) ");
     }
```
Fixed it.

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -419,265 +362,45 @@ private void buildWhereClauseForInsert(StringBuilder query) {
-          "`operation` int, `originalTransaction` bigint, `bucket` int, `rowId` bigint, "
-              + "`currentTransaction` bigint, `row` struct<");
+          "`operation` int, `originalTransaction` bigint, `bucket` int, `rowId` bigint, " + "`currentTransaction` bigint, `row` struct<");
+      List<String> columnDescs = getColumnDescs();
+      query.append(StringUtils.join(columnDescs, ','));
+      query.append(">) ");
```
Fixed it.

kuczoram commented:
Thanks a lot @ayushtkn for the review! I appreciate your comments very much!

I checked the comments from @aturoczy in the old PR. If I don't miss something, he had two suggestions, one about the style of an error message and one about writing unit tests. I fixed the error message and I added unit tests for the whole query generation. Do you think we should write unit tests for the individual methods as well? If you think that it would make sense, I can do that. But maybe I open a different Jira for that, so I won't block this one until it is finished.

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -275,137 +250,105 @@ String build() {
+      tblProperties.put("bucketing_version", String.valueOf(bucketingVersion));
+    }
+    if (sourceTab != null) { // to avoid NPEs, skip this part if sourceTab is null
+      for (Map.Entry<String, String> e : sourceTab.getParameters().entrySet()) {
+        if (e.getKey().startsWith("orc.")) {
+          tblProperties.put(e.getKey(), HiveStringUtils.escapeHiveCommand(e.getValue()));
+        }
+      }
```
Sure, I changed that.

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -419,265 +362,45 @@ private void buildWhereClauseForInsert(StringBuilder query) {
       query.append(
-          "`operation` int, `originalTransaction` bigint, `bucket` int, `rowId` bigint, "
-              + "`currentTransaction` bigint, `row` struct<");
+          "`operation` int, `originalTransaction` bigint, `bucket` int, `rowId` bigint, " + "`currentTransaction` bigint, `row` struct<");
```
Yeah, you are right. Fixed that.

kuczoram commented:
Thanks a lot @zratkai for the review, I appreciate it very much.

zratkai reviewed (changes requested):
Please check my answer.

kuczoram commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:
```diff
@@ -275,137 +250,105 @@ String build() {
+  protected void addTblProperties(StringBuilder query, Map<String, String> tblProperties) {
+    // add TBLPROPERTIES clause to query
+    boolean isFirst;
+    query.append(" TBLPROPERTIES (");
```
In practice, it should never happen that the tblProperties are empty, because the 'transactional=false' property is always added. But you are right, we should have a check here and not to add anything if the properties are empty.

kuczoram replied:
Fixed it.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderTestBase.java:
```diff
@@ -0,0 +1,143 @@
+
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.common.ValidCompactorWriteIdList;
+import org.apache.hadoop.hive.metastore.api.*;
```
Can we expand this import
```
import org.apache.hadoop.hive.metastore.api.FieldSchema;
import org.apache.hadoop.hive.metastore.api.Order;
import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
import org.apache.hadoop.hive.metastore.api.Table;
```

kuczoram replied:
Fixed it.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderTestBase.java:
```diff
@@ -0,0 +1,143 @@
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+import org.mockito.Mockito;
+
+import java.util.*;
```
can we expand these imports rather than using *
```
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
```

kuczoram replied:
Fixed it.

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForInsertOnly.java:
```diff
@@ -0,0 +1,274 @@
+      String storageHandler = sourceTab.getParameters().get(hive_metastoreConstants.META_TABLE_STORAGE);
+      if (storageHandler != null) {
+        String message =
+            "Table " + sourceTab.getTableName() + "has a storage handler (" + storageHandler + "). Failing compaction for this non-native table.";
```
There is a space missing before has, else if table names is tb123 it will print "tbl123has" a storage handler

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMmCompaction.java:
```diff
@@ -0,0 +1,472 @@
+    queryBuilder.setStorageDescriptor(storageDescriptor);
+    Exception exception = Assert.assertThrows(RuntimeException.class, queryBuilder::build);
+    String expectedMessage =
+        "Table comp_test_source_tablehas a storage handler (test_storage_handler). Failing compaction for this non-native table.";
```
table name and has should have space in b/w, once you correct the prod code, this test needs an update as well

kuczoram replied:
Fixed it.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMmCompaction.java:
```diff
@@ -0,0 +1,472 @@
+    queryBuilder.setSourceTab(sourceTable);
+    StorageDescriptor storageDescriptor = createStorageDescriptor();
+    queryBuilder.setStorageDescriptor(storageDescriptor);
+    Exception exception = Assert.assertThrows(RuntimeException.class, queryBuilder::build);
+    String expectedMessage =
+        "Table comp_test_source_tablehas a storage handler (test_storage_handler). Failing compaction for this non-native table.";
+    String actualMessage = exception.getMessage();
+    Assert.assertTrue(actualMessage.contains(expectedMessage));
```
Can we change this logic to something like
```
    String expectedMessage =
        "Table comp_test_source_tablehas a storage handler (test_storage_handler). Failing compaction for this non-native table.";
    Assert.assertThrows(expectedMessage, RuntimeException.class, queryBuilder::build);
```

& in other tests as well, if we are doing something similar

kuczoram replied:
Sure, fixed it.

ayushtkn commented on ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCompactionQueryBuilderForMmCompaction.java:
```diff
@@ -0,0 +1,472 @@
+    queryBuilder.setSourceTab(sourceTable);
+    StorageDescriptor storageDescriptor = createStorageDescriptor();
+    queryBuilder.setStorageDescriptor(storageDescriptor);
+    Exception exception = Assert.assertThrows(RuntimeException.class, queryBuilder::build);
+    String expectedMessage =
+        "Table comp_test_source_tablehas a storage handler (test_storage_handler). Failing compaction for this non-native table.";
+    String actualMessage = exception.getMessage();
+    Assert.assertTrue(actualMessage.contains(expectedMessage));
```
assertThrows can take the expected message, so maybe we can try that here, if possible

kuczoram replied:
Fixed it.

ayushtkn reviewed (commented):
Thanx @kuczoram, I have couple of minor comments, post that things LGTM

ayushtkn reviewed (approved):
LGTM
